### DISCUSSION TITLE
Extending System size and Correlation Function

### DISCUSSION THREAD

**Discussion by yonasG** [QuSpin/QuSpin#661] [General]
I am solving a spin Hamiltonian on a triangular lattice with C_3, inversion, and mirror-x Mx (but not My) symmetry. The interactions are bond-dependent so I construct the Hamiltonian on the a1, a2, and a3 directions independently. My goals are to calculate (i) the ground state energy E0 (ii) the correlation functions Cxx, Cyy, Czz and (iii) the static spin structure factors Sxx, Syy, Szz which are the spatial Fourier transforms of the correlation functions.

1) How can I modify the code to handle systems of size 6x5 or 6x6? On local machines, 6x5 and 6x6 systems keep restarting the Kernel so I suspect there's some memory overflow. How can I treat the Hamiltonian as a sparse matrix as shown [here]( https://github.com/QuSpin/QuSpin/discussions/630)

2) What is the best format to save the ground state wave function to use in the correlation function calculations? Even though the code below shows calculations of Psi_GS and correlation functions simultaneously, I plan to compute them separately for efficiency.

3) I don't have an alternative way to check so, when calculating the correlation functions is the definition/use of ss_coeff = [[1.0, i, j]] correct?

```
## General packages
import numpy as np

## QuSpin packages
from quspin.basis import spin_basis_general # spin basis constructor
from quspin.operators import hamiltonian

def mylinspace(start, stop, step=1.):
  """
  If start=1, stop=3, step=0.5, Output: array([1., 1.5, 2., 2.5, 3.])
  """
  return np.linspace(start, stop, int((stop - start) / step + 1))

def SitePosition(i,Ny):
    a1 = np.array([1.0, 0])
    a2 = np.array([0.5, 0.5*np.sqrt(3.0)])
    rx = int(np.floor(i/Ny)) # x-position on square lattice
    ry = i%Ny                # y-position on square lattice
    rVec = rx*a1 + ry*a2     
    return rVec
    

pi = np.pi

## System parameters
Nx, Ny = 4, 4       # linear dimensions of triangular lattice   
N=Nx*Ny
sites=np.arange(N)  # sites = [0,1,2,..., N-1]
x=sites%Nx          # x = [0,1,...,Nx, 0,1,...,Nx,...] (columns of lattice)
y=sites//Nx         # y = [0,0,...,0, 1,1,...,1,...] (rows of latice)

## Symmetries
Tx = (x+1)%Nx + Nx*y    # translation along x-direction
Ty = x + Nx*((y+1)%Ny)  # translation along y-direction
Px = x + Nx*(Ny-y-1)    # parity - reflection about x-axis
#Py = (Nx-x-1) + Nx*y    # parity - reflection about y-axis	
Z = -(sites+1)          # spin inversion symmetry

## set up Hamiltonian basis
basis_full = spin_basis_general(N, Nup=None, m=None, S='1/2', pauli=0)
#$ Symmetry reduced basis
basis_red=spin_basis_general(N,kxblock=(Tx,0),kyblock=(Ty,0),pxblock=(Px,0),zblock=(Z,0),pauli=0)
#
###### Exchange couplings ######
J1 = 1.0
Jone = str(round(J1,2)).replace('.', '_')
#
J2 = 1.0
Jtwo = str(round(J2,2)).replace('.', '_')
#
J3 = 3.0
Jthree = str(round(J3,2)).replace('.', '_')
#
h = 0.0
Bfld = '0'+ str(round(int(10*h),3))  #.replace('.', '_')
#h -> Bfld: 0.0 -> 00, 1.0 -> 010, 1.5 -> 015, 10.0 -> 0100, etc.
#
# Construct the Hamiltonian (Edited out)
#
static = [["xx",Hxx],["xy",Hxy],["xz",Hxz],\
		  ["yx",Hxy],["yy",Hyy],["yz",Hyz],\
		  ["zx",Hxz],["zy",Hyz],["zz",Hzz],["z",h_field]]    
dynamic =[]
#
# construct Hamiltonian and compute the g.s. energy and w.f.
H = hamiltonian(static, dynamic, basis=basis_red)
E,V= H.eigsh(k=1,which='SA',maxiter=1E4,return_eigenvectors=True)
print(f"E0 = {E[0]}")
#
psi_GS_red = V[:,0]   # ground state using the reduced basis
psi_GS_full= basis_red.project_from(psi_GS_red, sparse = False, pcon = True)
#
###### Correlation functions ###### 
Cxx, Cyy, Czz = (np.zeros((N,N),dtype=np.float64) for i in range(3))
for i in range(N):
    for j in range(N):
        ss_coeff = [[1.0, i, j]]
        SxSx_op = hamiltonian([['xx', ss_coeff]],[],basis=basis_full,check_herm=False, check_symm=False)
        SySy_op = hamiltonian([['yy', ss_coeff]],[],basis=basis_full,check_herm=False, check_symm=False)
        SzSz_op = hamiltonian([['zz', ss_coeff]],[],basis=basis_full,check_herm=False, check_symm=False)
        Cxx[i,j] = SxSx_op.expt_value(psi_GS_full).real
        Cyy[i,j] = SySy_op.expt_value(psi_GS_full).real
        Czz[i,j] = SzSz_op.expt_value(psi_GS_full).real
        
# np.savetxt('Czz_N'+str(N)+'_J1'+Jone+'_J3'+Jthree+'_h'+Bfld+'.txt',Czz)
#
######  Static spin Structure Factor (SSF)   ###### 
KxStep = (2.0*pi)/(10*Nx)
KxVals = mylinspace(-1.5*np.pi, 1.5*np.pi, KxStep)
#
KyStep = (2.0*pi)/(10*Ny)
KyVals = mylinspace(-1.5*np.pi, 1.5*np.pi, KyStep)
#
Sxx,Syy,Szz,SqTot  = (np.zeros((len(KxVals), len(KyVals))) for i in range(4))
for l, kx in enumerate(KxVals):
    for m, ky  in enumerate(KyVals):
        Qvec = (kx,ky)
        XX_sum, YY_sum, ZZ_sum = 0.0, 0.0, 0.0
        for i in range(N):
            for j in range(N):
                Rvec = SitePosition(i,Ny) - SitePosition(j,Ny)
                XX_sum += np.cos(np.dot(Qvec,Rvec)) * Cxx[i][j]
                YY_sum += np.cos(np.dot(Qvec,Rvec)) * Cyy[i][j]
                ZZ_sum += np.cos(np.dot(Qvec,Rvec)) * Czz[i][j]
        Sxx[l,m] = XX_sum/N
        Syy[l,m] = YY_sum/N
        Szz[l,m] = ZZ_sum/N
        SqTot[l,m] =  (XX_sum + YY_sum + ZZ_sum)/N
#
# np.savetxt('SqTot_N'+str(N)+'_J1'+Jone+'_J3'+Jthree+'_h'+Bfld+'.txt', SqTot)
```

**Comment by mgbukov** [comment]
1. you seem to be using the full basis to compute the correlation function; if you make use of symmetries, you can potentially reach larger system sizes. This [example](https://quspin.github.io/QuSpin/examples/example19.html#example19-label) from the online doc may give u some ideas. Ultimately the system sizes you can reach depend on your machine, and it is natural that you hit some limits. 

To answer your Q explicitly: any `hamiltonian` object is by default a sparse matrix, try printing it and you'll see.  

2. quspin doesn't really have any storing functions of its own; you can e.g., use [hdf5](https://docs.h5py.org/en/stable/)

3. you are likely not making use of any i <--> j symmetry (if present); other than that -- using the full basis is correct since local operators defined on sites (i,j) break translation symmetry. The only symmetry you can still use is total magnetization (Nup blocks) if present in your model. 

If you use the example mentioned above, then your current code can be your benchmark: of both codes agree you got it right with high probability; if they don't there's something wrong with one of them.

**Comment by yonasG** [comment]
Thank you for the reply!

Yonas
________________________________
